airbyte-integrations/connectors/source-google-search-console/source_google_search_console/streams.py

```diff
@@ -0,0 +1,303 @@
+            yield record
+
+    @property
+    def authenticator(self) -> Union[HttpAuthenticator, None]:
+        if self._client_id and self._client_secret and self._refresh_token:
+            return Oauth2Authenticator(
+                token_refresh_endpoint=GOOGLE_TOKEN_URI,
+                client_secret=self._client_secret,
+                client_id=self._client_id,
+                refresh_token=self._refresh_token,
+            )
+        elif self._service_account_info:
+            info = json.loads(self._service_account_info)
+            return ServiceAccountAuthenticator(service_account_info=info)
+
+
+class Sites(GoogleSearchConsole):
+    """
+    API docs: https://developers.google.com/webmaster-tools/search-console-api-original/v3/sites
+    """
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return f"/sites/{next_page_token or self._site_urls.pop(0)}"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield response.json()
+
+
+class Sitemaps(GoogleSearchConsole):
+    """
+    API docs: https://developers.google.com/webmaster-tools/search-console-api-original/v3/sitemaps
+    """
+
+    data_field = "sitemap"
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return f"/sites/{next_page_token or self._site_urls.pop(0)}/sitemaps"
+
+
+class SearchAnalytics(GoogleSearchConsole, ABC):
+    """
+    API docs: https://developers.google.com/webmaster-tools/search-console-api-original/v3/searchanalytics
+    """
+
+    data_field = "rows"
+    start_row = 0
+    value_field = None
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        refresh_token: str,
+        site_urls: list,
+        start_date: pendulum.datetime,
+        service_account_info: str,
+    ):
+        super().__init__(client_id, client_secret, refresh_token, site_urls, start_date, service_account_info)
+        self.search_types = ["web", "news", "image", "video"]
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return f"/sites/{self._site_urls[0]}/searchAnalytics/query"
+
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
+        return "date"
+
+    @property
+    def http_method(self) -> str:
+        return "POST"
+
+    def next_page_token(self, response: requests.Response) -> Optional[dict]:
+        result = {}
+
+        if len(response.json().get(self.data_field, [])) == ROW_LIMIT:
+            self.start_row += ROW_LIMIT
+
+        elif len(self.search_types) > 1:
```

Comment:
this should also use stream slices instead of pagination but you should be careful of how it interacts with checkpointing state